Add -debug-constraints-on-line flag

beccadax opened the pull request:
This new flag allows you to specify that you want constraint solver debug output for any expression with a given line number. It doesn't filter by file name; I think line number alone will usually be enough to make the `-debug-constraints` spew manageable. You can specify more than one `-debug-constraints-on-line` flag and it will print debug output for expressions on any of those lines. You can also specify line 0 to get debug output for expressions with invalid SourceLocs.

The implementation here saves and restores the property in the AST context's `LangOpts`. It would be nice to have an `isDebugMode` property in the constraint system instead, but plumbing that everywhere (especially into ASTDumper) is more work than I want to tackle right now.

beccadax committed:
Add -debug-constraints-on-line flag

brentdax commented:
@swift-ci please smoke test

brentdax commented:
@swift-ci please smoke test macOS platform

xedin commented:
@brentdax FYI, there is a `isDebugMode` already in `SolverStep` which currently checks `ASTContext::DebugConstraintSolver` but could be changed easily to do anything else.

brentdax commented:
@swift-ci please smoke test macOS platform

brentdax commented:
@xedin I did see that and actually started on the refactoring to move it into ConstraintSystem. But there are some places in e.g. ConstraintGraphNode that can't access it easily and ASTDumper decides whether to dump type variables by checking the LangOpts, so it rapidly turned into a larger diversion than I wanted to take while I'm working on something else.

harlanhaskins commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -339,6 +339,17 @@ static bool ParseLangArgs(LangOptions &Opts, ArgList &Args,
+    if (StringRef(A->getValue()).getAsInteger(10, line)) {
+      Diags.diagnose(SourceLoc(), diag::error_invalid_arg_value,
+                     A->getAsString(Args), A->getValue());
+      return true;
```
If there are gonna be multiple of these, would it make sense to return after the loop if it failed, instead of bailing on the first one?

brentdax replied:
This is actually an issue in the wider function, but I can correct that.

xedin commented on lib/Sema/CSSolver.cpp:
```diff
@@ -1020,11 +1020,30 @@ void ConstraintSystem::shrink(Expr *expr) {
+
+  unsigned line = 0;
+  if (expr->getLoc().isValid())
+    line = C.SourceMgr.getLineNumber(expr->getLoc());
```
I don't think this works when there is a single statement closure and line number points to its body e.g.

```foo(..) {
  _ = x + y <- this line
}
```

Since closure is type-checked together with enclosing context there is going to be no debug output. Also consider expressions with implicit containers (SanitizeExpr would mutate AST)? I think this method needs to evaluate whole AST to figure out if debug should be enabled or not.

brentdax replied:
My goal was to use the same line number `solveImpl()` logs when it starts solving an expression. I guess that should actually be `getStartLoc()`, though.

brentdax replied:
I've modified it to calculate both the start and end lines of the expression and activate debugging if any line in the list is between the two.

xedin replied:
Perfect, thank you!

beccadax committed:
Remove early returns in ParseLangArgs

Several places in language option parsing return early after diagnosing an error when they could instead continue parsing and discover addtional errors to diagnose. Change these to set an error flag instead.

OPT_package_description_version still has an early return on the theory that, if you’re using a too-old compiler to parse a Package.swift file, you’re probably confused about the Swift compiler version you’re using and further errors may be for arguments that would be accepted by a newer compiler.

brentdax commented:
@swift-ci please smoke test

brentdax commented:
@swift-ci please smoke test compiler performance
